README.md
# cafeyn-get
Ce script permet de récupérer un journal ou un magazine présent sur https://www.cafeyn.co dans la limite des capacités de notre compte existant.

Le but est de pouvoir lire un journal ou un magazine sur un support non compatible avec les applications fournies par Cafeyn. 
Il est évident que les fichiers ne doivent en aucun cas être conservées une fois la lecture terminée ou lorsque votre abonnement ne vous permet plus de les lire.


## Utilisation

### cafeyn_get
**Utilisation**  
```
usage: cafeyn_get.py [-h] [--no-clean] [--no-bookmark] [--output-folder OUTPUT_FOLDER] [--config CONFIG]
                     [--user-agent USER_AGENT] [--force] [--pause SECONDS]
                     [url]

Script pour sauvegarder une publication Cafeyn.

positional arguments:
  url                   L'URL de la publication à récupérer ou le chemin vers un fichier local contenant une liste
                        d'URLs

optional arguments:
  -h, --help            show this help message and exit
  --no-clean            Ne supprime pas le répertoire temporaire dans le cas où un PDF a été généré
  --no-bookmark         Ne met pas à jour le statut "lu" de la publication
  --output-folder OUTPUT_FOLDER, -o OUTPUT_FOLDER
                        Répertoire racine de téléchargement
  --config CONFIG       Fichier de configuration
  --user-agent USER_AGENT
                        User agent à utiliser
  --force               Ne demande pas la confirmation d'écrasement de fichier
  --pause SECONDS, -p SECONDS
                        Faire une pause (en secondes) entre chaque page
```

- L'URL doit ressembler à :
```
https://reader.cafeyn.co/{pays}/{identifiant_journal}/{identifiant_numero}
``` 

- Il est impératif d'avoir un fichier de configuration "cafeyn_get.cfg" dans lequel les 3 variables de cookies sont renseignées : 
```
cafeyn_authtoken = ...
cafeyn_webSessionId = ...
cafeyn_userGroup = ...
```
Pour les obtenir, identifiez vous sur https://www.cafeyn.co et recherchez vos cookies avec votre navigateur web.

#### Chrome  
Menu --> Plus d'outils --> Outils de développements  
Application / Storage / Cookies  
et recherchez le cookie "https://reader.cafeyn.co".  


#### Firefox  
Menu --> Developpement web --> Inspecteur de stockage --> Cookies  
et recherchez le cookie "https://reader.cafeyn.co".  



### cafeyn_get_selenium
**Utilisation**  
- Il faut tout d'abord lancer un navigateur Chrome en mode "debug". 
Pour cela, on peut exécuter le script : 
```
chrome_debug.bat
```
 
Ou adapter et exécuter cette ligne de commande : 
```
"C:\Program Files (x86)\Google\Chrome\Application\chrome.exe" --remote-debugging-port=9222 --user-data-dir="C:\temp\AutomationProfile"
```
  
- Il faut aller sur https://www.cafeyn.co s'identifier, choisir un journal, ouvrir le lecteur et se placer sur la première page. 
- L'URL doit ressembler à :
```
https://reader.cafeyn.co/fr/{identifiant_numero}/{identifiant_journal}
``` 
  
- On exécute le script Python : 
```
python cafeyn_get_selenium.py
```
  
- Les fichiers seront enregistrés dans le répertoire de téléchargement. 


## Installation
### Prérequis
- Python 3.7+ (non testé avec les versions précédentes)
- pip
- Librairies SSL
- Drivers Chrome

#### Sous Windows
##### Python
Allez sur ce site :  
https://www.python.org/downloads/windows/  
et suivez les instructions d'installation de Python 3.

##### Pip
- Téléchargez [get-pip.py](https://bootstrap.pypa.io/get-pip.py) dans un répertoire.
- Ouvrez une ligne de commande et mettez vous dans ce répertoire.
- Entrez la commande suivante :  
```
python get-pip.py
```
- Voilà ! Pip est installé !
- Vous pouvez vérifier en tapant la commande :  
```
pip -v
```

##### Librairies SSL
- Vous pouvez essayer de les installer avec la commande :  
```
pip install pyopenssl
```
- Vous pouvez télécharger [OpenSSL pour Windows](http://gnuwin32.sourceforge.net/packages/openssl.htm). 

##### Drivers Chrome
- Il faut télécharger le fichier "chromedriver.exe" [sur le site de Chromium](https://chromedriver.chromium.org/downloads) et le copier dans le répertoire 
```
bin\
```
- Il est possible de le renommer en "chromedriverXX.exe" où "XX" est le numéro de la version. Le système ira chercher celui qui convient à votre version de Chrome installée.


#### Sous Linux
Si vous êtes sous Linux, vous n'avez pas besoin de moi pour installer Python, Pip ou SSL...  

### Téléchargement
- Vous pouvez cloner le repo git :  
```
git clone https://github.com/izneo-get/cafeyn-get.git
```
ou  
- Vous pouvez télécharger uniquement le binaire Windows (expérimental).  


### Configuration
(pour la version "script" uniquement)
```
pip install -r requirements.txt
```


changelog.md
# Changelog

## cafeyn_get.py 

### Version 0.06.3 (2021-03-13)
- [BUGFIX] Adaptation à la nouvelle façon d'appeler les pages. 
- [NEW] Vérification de la version au lancement de l'application. 

### Version 0.06.2 (2021-03-07)
- [CHANGE] L'URL peut contenir un autre code pays que "fr". 

### Version 0.06.1 (2021-03-06)
- [NEW] Ajout de l'option `--pause SECONDS` qui permet de faire une pause entre chaque récupération de page. 

### Version 0.06.0 (2020-10-21)
- [NEW] Mise à jour du bookmark pour passer la publication en "lu". 
- [NEW] Ajout du paramètre "--no-bookmark" pour ne pas mettre à jour le bookmark. 

### Version 0.05.0 (2020-10-18)
- [NEW] Compatibilité avec les anciennes publications. 

### Version 0.04.0 (2020-10-18)
- [CHANGE] Utilisation d'une nouvelle clé RSA à chaque appel. 

### Version 0.03.1 (2020-10-09)
- [NEW] Si aucune URL passée en paramètre, on la demande. Cela permet d'exécuter le programme sans passer par une ligne de commande. 
- [NEW] Avertissement quand la publication n'est pas dans l'abonnement. 
- [NEW] Alerte si le fichier de destination existe déjà. 

### Version 0.02.0 (2020-10-08)
- [NEW] Utilisation de la date de sortie de la publication dans le nom du fichier. 

### Version 0.01.0 (2020-10-08)
- [NEW] Version initiale. 
- [NEW] Récupération des PDF en direct. 


## cafeyn_get_selenium.py 

### Version 0.01.0 (2020-10-06)
- [NEW] Ajout d'un paramètre "--no-pdf" pour ne pas créer de PDF. 
- [NEW] Ajout d'un paramètre "--no-clean" pour ne pas nettoyer les répertoire d'image après avoir généré un PDF. 
- [NEW] Ajout d'un paramètre "--image-format" (ou "-f") pour choisir le format des images ("jpg" ou "png"). 
- [NEW] Ajout d'un paramètre "--output-folder" (ou "-o") pour choisir le répertoire racine de téléchargement. 
- [CHANGE] Recherche du driver Chrome de manière récursive. 
- [BUGFIX] Détection des images vides. 

### Version 0.01.0 (2020-10-06) 
- [NEW] Version initiale. 
- [NEW] Récupération des images en pilotant un navigateur Chrome. 


cafeyn_get_selenium.py
# -*- coding: utf-8 -*-
__version__ = "0.02.0"
"""
Source : https://github.com/izneo-get/cafeyn-get

Ce script permet de récupérer un journal ou un magazine présent sur https://www.cafeyn.co dans la limite des capacités de notre compte existant.
"""
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.chrome.options import Options
import argparse
import configparser
import re
import sys
import os
import glob
from io import BytesIO
from PIL import Image, ImageOps
import base64
import time
import img2pdf
import shutil


def get_file_content_chrome(driver, uri):
    result = driver.execute_async_script(
        """
    var uri = arguments[0];
    var callback = arguments[1];
    var toBase64 = function(buffer){for(var r,n=new Uint8Array(buffer),t=n.length,a=new Uint8Array(4*Math.ceil(t/3)),i=new Uint8Array(64),o=0,c=0;64>c;++c)i[c]="ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/".charCodeAt(c);for(c=0;t-t%3>c;c+=3,o+=4)r=n[c]<<16|n[c+1]<<8|n[c+2],a[o]=i[r>>18],a[o+1]=i[r>>12&63],a[o+2]=i[r>>6&63],a[o+3]=i[63&r];return t%3===1?(r=n[t-1],a[o]=i[r>>2],a[o+1]=i[r<<4&63],a[o+2]=61,a[o+3]=61):t%3===2&&(r=(n[t-2]<<8)+n[t-1],a[o]=i[r>>10],a[o+1]=i[r>>4&63],a[o+2]=i[r<<2&63],a[o+3]=61),new TextDecoder("ascii").decode(a)};
    var xhr = new XMLHttpRequest();
    xhr.responseType = 'arraybuffer';
    xhr.onload = function(){ callback(toBase64(xhr.response)) };
    xhr.onerror = function(){ callback(xhr.status) };
    xhr.open('GET', uri);
    xhr.send();
    """,
        uri,
    )
    if type(result) == int:
        raise Exception("Request failed with status %s" % result)
    return base64.b64decode(result)


def clean_name(name):
    """Permet de supprimer les caractères interdits dans les chemins.

    Parameters
    ----------
    name : str
        La chaine de caractère d'entrée.

    Returns
    -------
    str
        La chaine purgée des tous les caractères non désirés.
    """
    chars = '\\/:*<>?"|'
    for c in chars:
        name = name.replace(c, "_")
    name = re.sub(r"\s+", " ", name)
    name = re.sub(r"\.+$", "", name)
    return name

def is_only_white(im):
    """Permet de supprimer les bords blancs. 

    Parameters
    ----------
    im : Image
        L'image que l'on souhaite traiter.

    Returns
    -------
    Image
        L'image sans les bords blancs.
    """
    tmp_im = im.convert("RGB")
    tmp_im = ImageOps.invert(tmp_im)
    bbox = tmp_im.getbbox()
    if not bbox:
        return True
    else:
        return False

if __name__ == "__main__":
    
    # Parse des arguments passés en ligne de commande.
    parser = argparse.ArgumentParser(
        description="""Ce script permet de récupérer un journal ou un magazine présent sur https://www.cafeyn.co dans la limite des capacités de notre compte existant."""
    )
    parser.add_argument(
        "--no-pdf",
        action="store_true",
        default=False,
        help="Ne génère pas de PDF.",
    )
    parser.add_argument(
        "--no-clean",
        action="store_true",
        default=False,
        help="Ne supprime pas le répertoire temporaire dans le cas où un PDF a été généré.",
    )
    parser.add_argument(
        "--image-format",
        "-f",
        choices={"png", "jpg"},
        type=str,
        default="jpg",
        help="Format des images.",
    )
    parser.add_argument(
        "--output-folder",
        "-o",
        type=str,
        default=None,
        help="Répertoire racine de téléchargement",
    )

    args = parser.parse_args()

    to_pdf = not args.no_pdf
    no_clean = args.no_clean
    output_format = args.image_format

    config = configparser.ConfigParser()
    config_file = re.sub(r"(.+)\.(.+)", r"\1", sys.argv[0]) + ".ini"
    prefered_driver = "./bin/chromedriver.exe"
    if os.path.isfile(config_file):
        config.read(config_file, encoding="utf-8")
        prefered_driver = config.get(
            "DEFAULT", "prefered_driver", fallback=prefered_driver
        )

    # On se rattache a une instance de Chrome existante
    chrome_options = Options()
    chrome_options.add_experimental_option("debuggerAddress", "127.0.0.1:9222")
    chrome_driver = prefered_driver
    try:
        driver = webdriver.Chrome(chrome_driver, options=chrome_options)
    except:
        # Ce driver n'est pas compatible.
        print(f'Impossible de se connecter avec le driver "{chrome_driver}"')
        for filename in glob.iglob("./**/chromedriver*.exe", recursive=True):
            chrome_driver = filename
            try:
                driver = webdriver.Chrome(chrome_driver, options=chrome_options)
            except:
                print(f'Impossible de se connecter avec le driver "{chrome_driver}"')
                continue
            prefered_driver = filename
            break

    config["DEFAULT"]["prefered_driver"] = prefered_driver
    with open(config_file, "w") as configfile:
        config.write(configfile)

    body = driver.find_element_by_tag_name("body")
    title = (
        driver.find_element_by_tag_name("h1").text
        + " ("
        + driver.current_url.replace("https://reader.cafeyn.co/fr/", "")
        + ")"
    )

    output_folder = os.path.dirname(os.path.abspath(sys.argv[0])) + "/DOWNLOADS"
    if args.output_folder:
        output_folder = args.output_folder
    
    if not os.path.exists(output_folder):
        os.mkdir(output_folder)

    save_path = output_folder + "/" + clean_name(title)
    if not os.path.exists(save_path):
        os.mkdir(save_path)

    termine = 0
    pages_done = []
    while termine <= 3:
        canvas = driver.execute_script(
            "return document.getElementsByTagName('canvas');"
        )
        new_pages = 0
        for c in canvas:
            parent_element = driver.execute_script(
                "return arguments[0].parentElement.id",
                c,
            )
            if parent_element not in pages_done:
                loaded = False
                attempts = 20
                while not loaded and attempts > 0:
                    try:
                        blob_url = driver.execute_script(
                            """
                        const blob = await new Promise(resolve => arguments[0].toBlob(resolve));
                        return URL.createObjectURL(blob);
                        """,
                            c,
                        )
                    except:
                        attempts = attempts - 1
                        continue
                    bytes = get_file_content_chrome(driver, blob_url)
                    if len(bytes) > 20000:
                        im = Image.open(BytesIO(bytes))
                        # Pour vérifier si ce n'est pas une image entièrement blanche, on la crop.
                        if not is_only_white(im):
                            loaded = True
                if loaded:
                    filename = ("00000" + parent_element.replace("canvas", ""))[-3:]
                    if output_format == 'jpg':
                        im = im.convert("RGB")
                        im.save(f"{save_path}/{filename}.{output_format}", quality=95)
                    else:
                        im.save(f"{save_path}/{filename}.{output_format}")
                    pages_done.append(parent_element)
                    new_pages = new_pages + 1
                    driver.execute_script(f"URL.revokeObjectURL('{blob_url}');")
                else:
                    print("Waiting")
                    attempts = attempts - 1
                    time.sleep(1)
        if new_pages == 0:
            termine = termine + 1
        else:
            termine = 0
        body.send_keys(Keys.ARROW_RIGHT)
        print(".", end="", flush=True)
        time.sleep(1)

    print("\nToutes les pages sont récupérées.")
    
    if to_pdf:
        print("Création du PDF...")
        with open(output_folder + "/" + clean_name(title) + ".pdf", "wb") as f:
            imgs = []
            for fname in os.listdir(save_path):
                if not fname.endswith(".jpg"):
                    continue
                path = os.path.join(save_path, fname)
                if os.path.isdir(path):
                    continue
                imgs.append(path)
            f.write(img2pdf.convert(imgs))

        if not no_clean:
            shutil.rmtree(save_path)
        print("OK")
        

cafeyn_get.py
# -*- coding: utf-8 -*-
__version__ = "0.06.3"
"""
Source : https://github.com/izneo-get/cafeyn-get
Date : 2021-03-13
"""

import requests
from requests.adapters import HTTPAdapter
from urllib3.util import Retry
import re
import os
import sys
import html
import argparse
import configparser
import shutil
import time
# from bs4 import BeautifulSoup
from PIL import Image
import json
from Crypto.PublicKey import RSA
from Crypto.Cipher import AES, PKCS1_OAEP
from Crypto.Util import Padding
import base64
from datetime import datetime
import math
from PyPDF2 import PdfFileMerger
import os.path
from os import path
import time


def requests_retry_session(
    retries=3,
    backoff_factor=1,
    status_forcelist=(500, 502, 504),
    session=None,
):
    """Permet de gérer les cas simples de problèmes de connexions."""
    session = session or requests.Session()
    retry = Retry(
        total=retries,
        read=retries,
        connect=retries,
        backoff_factor=backoff_factor,
        status_forcelist=status_forcelist,
    )
    adapter = HTTPAdapter(max_retries=retry)
    session.mount("http://", adapter)
    session.mount("https://", adapter)
    return session


def clean_name(name):
    """Permet de supprimer les caractères interdits dans les chemins.

    Parameters
    ----------
    name : str
        La chaine de caractère d'entrée.

    Returns
    -------
    str
        La chaine purgée des tous les caractères non désirés.
    """
    chars = '\\/:*<>?"|'
    for c in chars:
        name = name.replace(c, "_")
    name = re.sub(r"\s+", " ", name)
    name = re.sub(r"\.+$", "", name)
    return name

def check_version():
    latest_version_url = 'https://raw.githubusercontent.com/izneo-get/cafeyn-get/master/VERSION'
    res = requests.get(latest_version_url)
    if res.status_code != 200:
        print(f"Version {__version__} (impossible to check official version)")
    else:
        latest_version = res.text.strip()
        if latest_version == __version__:
            print(f"Version {__version__} (official version)")
        else:
            print(f"Version {__version__} (official version is different: {latest_version})")
            print("Please check https://github.com/izneo-get/cafeyn-get/releases/latest")
    print()

if __name__ == "__main__":
    # Parse des arguments passés en ligne de commande.
    parser = argparse.ArgumentParser(
        description="""Script pour sauvegarder une publication Cafeyn."""
    )
    parser.add_argument(
        "url",
        type=str,
        default="",
        nargs="?",
        help="L'URL de la publication à récupérer ou le chemin vers un fichier local contenant une liste d'URLs",
    )
    parser.add_argument(
        "--no-clean",
        action="store_true",
        default=False,
        help="Ne supprime pas le répertoire temporaire dans le cas où un PDF a été généré",
    )
    parser.add_argument(
        "--no-bookmark",
        action="store_true",
        default=False,
        help="Ne met pas à jour le statut \"lu\" de la publication",
    )
    parser.add_argument(
        "--output-folder",
        "-o",
        type=str,
        default=None,
        help="Répertoire racine de téléchargement",
    )
    parser.add_argument(
        "--config", type=str, default=None, help="Fichier de configuration"
    )
    parser.add_argument(
        "--user-agent", type=str, default=None, help="User agent à utiliser"
    )
    parser.add_argument(
        "--force",
        action="store_true",
        default=False,
        help="Ne demande pas la confirmation d'écrasement de fichier",
    )
    parser.add_argument(
        "--pause",
        "-p",
        type=float,
        metavar="SECONDS",
        default=0,
        help="Faire une pause (en secondes) entre chaque page",
    )

    # Vérification que c'est la dernière version.
    check_version() 

    args = parser.parse_args()
    no_clean = args.no_clean
    no_bookmark = args.no_bookmark
    force_overwrite = args.force
    pause_sec = args.pause

    # Lecture de la config.
    config = configparser.RawConfigParser()
    if args.config:
        config_name = args.config
    else:
        config_name = re.sub(r"\.py$", ".cfg", os.path.abspath(sys.argv[0]))
        config_name = re.sub(r"\.exe$", ".cfg", config_name)
    config.read(config_name)

    def get_param_or_default(config, param_name, default_value, cli_value=None):
        if cli_value is None:
            return (
                config.get("DEFAULT", param_name)
                if config.has_option("DEFAULT", param_name)
                else default_value
            )
        else:
            return cli_value

    output_folder = get_param_or_default(
        config,
        "output_folder",
        os.path.dirname(os.path.abspath(sys.argv[0])) + "/DOWNLOADS",
        args.output_folder,
    )
    if not os.path.exists(output_folder):
        os.mkdir(output_folder)

    user_agent = get_param_or_default(config, "user_agent", "", args.user_agent)
    # Paramétrages liés à la session.
    cafeyn_authtoken = get_param_or_default(config, "cafeyn_authtoken", "")
    cafeyn_webSessionId = get_param_or_default(config, "cafeyn_webSessionId", "")
    cafeyn_userGroup = get_param_or_default(config, "cafeyn_userGroup", "")

    base_url = args.url
    while base_url.upper() != "Q" and not re.match(
        "https://reader.cafeyn.co/(.+?)/(.+?)/(.+)", base_url
    ):
        base_url = input(
            'URL de la publication au format "https://reader.cafeyn.co/{pays}/{publicationId}/{issueId}" ("Q" pour quitter) : '
        )

    if base_url.upper() == "Q":
        sys.exit()

    if not cafeyn_authtoken or not cafeyn_webSessionId or not cafeyn_userGroup:
        print("[ERREUR] Impossible de trouver les valeurs dans la configuration.")
        print(f'cafeyn_authtoken = "{cafeyn_authtoken}"')
        print(f'cafeyn_webSessionId = "{cafeyn_webSessionId}"')
        print(f'cafeyn_userGroup = "{cafeyn_userGroup}"')

    if not user_agent:
        user_agent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36"
    s = requests.Session()

    _, publication, issue = re.match("https://reader.cafeyn.co/(.+?)/(.+?)/(.+)", base_url).groups()

    mag_infos_url = f"https://api.lekiosk.com/api/v1/reader/publications/{publication}/issues/{issue}/signedurls"

    # On génère une nouvelle clé RSA.
    # priv_key = RSA.import_key(open('privatekey.pkcs8').read())
    priv_key = RSA.generate(1024)
    pub_key = priv_key.publickey().exportKey("PEM").decode("utf-8")
    for e in ["-----BEGIN PUBLIC KEY-----", "-----END PUBLIC KEY-----", "\n"]:
        pub_key = pub_key.replace(e, "")

    baseValue = pub_key

    headers = {
        "authority": "api.lekiosk.com",
        "pragma": "no-cache",
        "cache-control": "no-cache",
        "accept": f"accesskeyid:123456;isretina:false;appversion:5.0.0;content-type:application/json;cookieprefix:Cafeyn_;appId:canal;webSessionId:{cafeyn_webSessionId};authtoken:{cafeyn_authtoken};baseValue:{baseValue}",
        "user-agent": user_agent,
        "x-usergroup": cafeyn_userGroup,
        "origin": "https://reader.cafeyn.co",
        "sec-fetch-site": "cross-site",
        "sec-fetch-mode": "cors",
        "sec-fetch-dest": "empty",
        "referer": "https://reader.cafeyn.co/",
        "accept-language": "fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7",
        "dnt": "1",
    }

    timestamp = math.floor(datetime.timestamp(datetime.now()))
    params = (("timestamp", f"{timestamp}000"),)

    r = requests_retry_session(session=s).get(
        mag_infos_url,
        # cookies=s.cookies,
        allow_redirects=True,
        headers=headers,
        params=params,
    )
    mag_infos = json.loads(r.text)["result"]

    if mag_infos["isPurchased"] == False:
        print(
            "[WARNING] Cette publication n'est pas disponible dans votre abonnement. Vous n'aurez que les pages gratuites. "
        )

    enc_session_key = base64.b64decode(mag_infos["baseValue"])

    cipher = PKCS1_OAEP.new(priv_key)
    cipher_dec = cipher.decrypt(enc_session_key)

    title = clean_name(mag_infos["title"])
    issueNumber = clean_name(str(mag_infos["issueNumber"]))
    releaseDate = clean_name(str(mag_infos["releaseDate"]))
    if len(releaseDate) > 10:
        releaseDate = " (" + releaseDate[0:10] + ")"
    else:
        releaseDate = ""

    full_pdf_path = (
        output_folder
        + "/"
        + clean_name(title + " " + issueNumber + releaseDate)
        + ".pdf"
    )

    answer = ""
    if force_overwrite:
        answer = "O"
    while path.exists(full_pdf_path) and answer not in ["O", "Y", "N", "Q"]:
        answer = input(
            f'Le fichier de destination "{full_pdf_path}" existe déjà. Voulez-vous l\'écraser ? [O]ui / [N]on : '
        ).upper()

    if answer.upper() in ["N", "Q"]:
        sys.exit()

    done = 0
    total = len(mag_infos["signedUrls"])
    print(f"[{done} / {total}]", end="\r", flush=True)
    max_page = 0
    for page in mag_infos["signedUrls"]:
        headers = {
            'User-Agent': user_agent,
            'Accept': 'application/json, text/plain, */*',
            'Accept-Language': 'fr,fr-FR;q=0.8,en-US;q=0.5,en;q=0.3',
            'AppId': 'lk',
            'Origin': 'https://reader.cafeyn.co',
            'Connection': 'keep-alive',
            'Referer': base_url,
            'Pragma': 'no-cache',
            'Cache-Control': 'no-cache',
            'TE': 'Trailers',
            }
        r = requests_retry_session(session=s).get(
            mag_infos["signedUrls"][page]["pdfUrl"],
            # cookies=s.cookies,
            allow_redirects=True,
            headers=headers,
        )
        if r.status_code != 200:
            print(f"Impossible de récupérer la page {page}")
            break
        page_str = ("000" + page)[-3:]
        # f = open(f"{title} {issueNumber} - {page_str}.bin", "wb")
        # f.write(r.content)
        # f.close()

        save_path = output_folder + "/" + title + "_" + issueNumber
        if not os.path.exists(save_path):
            os.mkdir(save_path)

        if re.fullmatch(r"^[0-9a-fA-F]+$", cipher_dec.decode()) is not None:
            # Nouveau format
            e = mag_infos["issueId"] * ord("#")
            t = mag_infos["publicationId"] * ord("*")
            b = (hex(t)[2:].upper() + "00000000000")[0:10]
            a = (hex(e)[2:].upper() + "00000000000")[0:10]
            key = b + cipher_dec.decode() + a
            iv = key[0:16]
        else:
            # Ancien format
            key = (str(mag_infos["issueId"]) + "-" + page + cipher_dec.decode())[0:16]
            # iv = ''
            # for k in key:
            #     iv = iv + hex(ord(k))[2:]
            iv = key

        key = key.encode("utf-8")
        iv = iv.encode("utf-8")

        aes = AES.new(key, AES.MODE_CBC, iv)
        # decrypted_content = aes.decrypt(r.content)
        decrypted_content = Padding.unpad(aes.decrypt(r.content), AES.block_size)

        f = open(f"{save_path}/{title} {issueNumber} - {page_str}.pdf", "wb")
        f.write(decrypted_content)
        f.close()
        done = done + 1
        max_page = int(page) if int(page) > max_page else max_page
        print(f"[{done} / {total}]", end="\r", flush=True)
        time.sleep(pause_sec)

    print("\nCompilation du PDF...")

    pdfs = []
    for fname in os.listdir(save_path):
        if not fname.endswith(".pdf"):
            continue
        path = os.path.join(save_path, fname)
        if os.path.isdir(path):
            continue
        pdfs.append(path)

    merger = PdfFileMerger()
    for pdf in pdfs:
        merger.append(pdf)
    merger.write(full_pdf_path)
    merger.close()

    if not no_clean:
        shutil.rmtree(save_path)

    # On indique à Cafeyn que le magazine est lu.
    if not no_bookmark:
        print("Mise à jour du bookmark...")
        headers = {
            "authority": "api.lekiosk.com",
            "pragma": "no-cache",
            "cache-control": "no-cache",
            "accept": f"accesskeyid:123456;isretina:false;appversion:5.0.0;content-type:application/json;cookieprefix:Cafeyn_;appId:canal;webSessionId:{cafeyn_webSessionId};authtoken:{cafeyn_authtoken};baseValue:{baseValue}",
            "user-agent": user_agent,
            "x-usergroup": cafeyn_userGroup,
            'content-type': 'application/json;charset=UTF-8',
            "origin": "https://reader.cafeyn.co",
            "sec-fetch-site": "cross-site",
            "sec-fetch-mode": "cors",
            "sec-fetch-dest": "empty",
            "referer": base_url,
            "accept-language": "fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7",
            "dnt": "1",
        }

        data = (
            '[{"issueId":'
            + str(mag_infos["issueId"])
            + ',"lastReadPage":'
            + str(max_page)
            + ',"lastReadMode":0}]'
        )
        r = requests_retry_session(session=s).put(
            "https://api.lekiosk.com/api/v1/users/me/readings/issues",
            # cookies=s.cookies,
            allow_redirects=True,
            headers=headers,
            params=params,
            data=data,
        )
        if r.status_code != 200:
            print("[WARNING] Impossible de mettre à jour le marque page.")

    print("Terminé !")
